### ISSUE TITLE
Should().BeEquivalentTo that allows mapping

### ISSUE THREAD

**Issue by fredgate** [fluentassertions/fluentassertions#535]
Suppose that I have an entity class `Order` and a DTO class `OrderDto` like this :
```csharp
class Order
{
  public long Id { get; set; }
  public DateTime Date { get; set; }
  public int State { get; set; }
}

class OrderDto
{
  public long Id { get; set; }
  public DateTime Date { get; set; }
  public StateDto State { get; set; }
}

class StateDto
{
  public int Code { get; set; }
  public String Text { get; set; }
}
```

Now, I have a test where I want to verify that a retrieved DTO is equivalent to existing entity.
```csharp
Order existing;
OrderDto actual;

// Arrange
existing = ....;

// Act
actual = ....;

// Assert
actual.ShouldBeEquivalentTo(existing);
```

This assertion fails because there is no `State.Code` and `State.Text` properties in the `Order` class.
So I changed my assertion to :
```csharp
actual.ShouldBeEquivalentTo(existing, o => o.ExcludingMissingMembers());
```
Now it works but it does not assert that the `State.Code` nested property of the `OrderDto` class is equal to the `State` property of the `Order` class.
So my test is not complete.
What would be the best way to assert the state code of the retrieved DTO ? (I have other similar properties in the DTO class).

**Comment by dennisdoomen** [comment]
Unfortunately you would need another separate statement. I would still like to introduce an option to explicitly tell the API how to compare two different properties.

**Comment by fredgate** [comment]
Thank you for the feedback. That's what I thought because I could not find anything in the API.
Maybe it would be possible to configure this by adding a method to IEquivalencyAssertionOptions as for example :
```csharp
actual.ShouldBeEquivalentTo(
  existing,
  o => o.ExcludingMissingMembers().Map(x => x.State.Code, x => x.State));
```

**Comment by dennisdoomen** [comment]
Exactly what I had in mind. But I'm a bit preoccupied with FA 5.0 first.

**Comment by nohwnd** [comment]
I would really like to have this feature, but I never developed anything for FluentAssertions, so please help me review if the general idea is at least in the same ballpark, before I start implementing a prototype. :)

First we need the Map to take two different parameter type s and e, (`TSubject s` for subject and `TExpectation e` for expected). That might be a slight problem because the extension only takes TExpectation from expected. The type of subject is lost because ObjectAssertion uses it as `object` instead of `T`. But let's not think about that for now, there might be some repetition of types, but I can live with that for now.

Using the example from the previous post, say we are flattening a domain object to a simpler dto, and in our converter we map from `Address.State.Code` to `AddressDto.StateCode`.

```csharp
var subject = Mapper.FromDto(expected);
subject.ShouldBeEquivalentTo(expected, o => o.ExcludingMissingMembers()
    .Map(s => s.StateCode, e => e.State.Code));
```

- What is the correct order to use here? subject -> expected or expected -> subject? To me subject -> expected seems more natural.

I inspected the library code quickly and tried debugging it a bit and from what I understood the minimum I need to make a prototype is:

- class similar to IncludeMemberByPredicateSelectionRule where I will specify the maping 
- extend EquivalencyAssertionOptions to add the selection rule same way it does with Include
- update MustMatchByNameRule to check Selection rules in the config parameter. It will look there to see if any "mapped" selection rules are applied read the mapping info from and find the property on the subject based on the map from the current expectedMember.

- I also saw there are UserEquivalencySteps, do you think I might leverage those somehow to implement this outside of the library, so I can use it right away, or is that for something else?

**Comment by dennisdoomen** [comment]
Hi Jakub,

Thanks for showing interest for this feature. It's something that I wanted for a long time, but my hands are tight to #588 right now.

> That might be a slight problem because the extension only takes TExpectation from expected. 

You'll most definitely need to specify the type of the subject:

`actual.ShouldBeEquivalentTo(existing, o => o.Map<TSubject>(s => s.State, e => e.State.Code));`

But that can become quite ugly fast if you want to map more than one property. Maybe we can create an additional method to set the subject type:

`actual.ShouldBeEquivalentTo(existing, o => o.WithSubject<TSubject>().Map(s => s.State, e => e.State.Code));`

> To me subject -> expected seems more natural.

I agree.

>  so I can use it right away, or is that for something else?

You could do both. Implement it properly a PR for FA 5.0 and in the mean time add it as a user-equivalency-step. 

>  inspected the library code quickly and tried debugging it a bit and from what I understood the minimum I need to make a prototype is:

`IncludeMemberByPredicateSelectionRule` will override the inclusion of all properties. Our new `Map` method should not affect the inclusion or exclusion of members in any way. It's more like an `IMemberMatchingRule`

**Comment by nohwnd** [comment]
Thanks for the quick reply, I am glad I was not totally off. I also have to say that the code is way easier to understand than I expected. :) 

>  Our new Map method should not affect the inclusion or exclusion of members in any way.

You are right, I was thinking about the Map as doing Map & Include, because mapping an Excluded member seems to be pointless, but that would limit the compared members to just the ones that we mapped which is not what we want.

So expanding on the example you gave, to Include just a single member and map it we would have to do: 

```csharp
actual.Should().BeEquivalentTo(existing, 
    o => o
        .Including(e=>e.State.Code)
        .WithSubject<OrderDto>().Map(s => s.State, e => e.State.Code)
);
```

_Note: looks like `actual.ShouldBeEquivalentTo()` is no longer valid for object assertion, so the code would be `actual.Should().BeEquivalentTo()`_

**Comment by dennisdoomen** [comment]
> Note: looks like actual.ShouldBeEquivalentTo() is no longer valid for object assertion, so the code would be actual.Should().BeEquivalentTo()

Yes, those are the breaking changes for 5.0, for which this feature is planned.

**Comment by nohwnd** [comment]
Just few notes from what I found so far, so I don't forget, and you have some chance to steer my thinking :) 

* It would be better to call the helper method `WithSubjectType<T>` to avoid suggesting that we can change the actual subject, because we need second helper method `WithSubjectTypeFrom<T>(T instance)` to allow mapping subjects that have anonymous type. In this method we only use the instance to infer the type, and nothing else. Setting the instance paramether to be optional is afaik not possible because there is no default value that is assignable to any T, and would make the api more confusing anyway.

```csharp
subject.Should().
    BeEquivalentTo(other, options => options
       .Including(p=>p.FirstName));
       .WithSubjectTypeFrom(subject)
       .Map(s=>s.Name, o => o.FirstName)
```

* Currently I am producing `SubjectSpecification<TExpectation, TSubject>` from the `WithSubject` and `Map` methods. SubjectSpecification holds on to  the instance of `EquivalencyAssertionOptions<TExpectation>` that produced it and implicitly casts to it so we can terminate the options successfullly. This only allows the Mapping to be specified as last in the options, but it's the least invasive approach I could think of. Another approach would be to make  `EquivalencyAssertionOptions` take two type parameters and specialize that to <TExpectation,object> for the usual use. That would allow to the Maps and Includes (and all other) to be mixed as desired, but I want to have working proof of concept with minimal changes in the code and some minor restrictions, before thinking about the api and so on. 

```csharp
public class SubjectSpecification<TExpectation, TSubject>
{
    private readonly EquivalencyAssertionOptions<TExpectation> assertionOptions;

    public SubjectSpecification(EquivalencyAssertionOptions<TExpectation> assertionOptions)
    {
        this.assertionOptions = assertionOptions;
    }

    public SubjectSpecification<TExpectation, TSubject> Map(
        Expression<Func<TSubject, object>> subjectExpression,
        Expression<Func<TExpectation, object>> expectationExpression
    ){
        assertionOptions.AddMappingRule(new MappingRule(...));
        return this;
    }

    public static implicit operator EquivalencyAssertionOptions<TExpectation>(
        SubjectSpecification<TExpectation, TSubject> instance
    ){
        return instance.assertionOptions;
    }
}
```

* Contrary to what we agreed, we do need to change the member selection a bit becasue of `ExcludeMissingMembers`. Given the following anonymous objects:

```csharp
// subject
new { State = 2 }
// expectation
new { Code = 1 }
```

Comparing them without any mapping the assertion would pass because there are no matching members.

With mapping `State -> Code`, the assertion should fail because the Code member is mapped to State and their values don't match.

And it get's more complicated with nested mappings and multiple mappings, behavior of which I am currently figuring out, but I think I will need to change how members are selected. Originally I thought the mapping can work by running in a match rule but that seems impossible now. Say for example I would run in context of `MustMatchByNameRule` ( which either finds corresponding member on subject, or returns null), and I am given mapping `State -> HomeState.Code` & `StateName -> HomeState.Name`. Normally I would get `HomeState` expectation member, find it on the subject, and move forward to next members and child members. But with mapping, I should select something to both avoid terminating the assertion with missing member error, and to be able to progress to child members of HomeState, but I cannot select any member on subject because nothing on the subject should match the `HomeState` member on expectation. 

As I say I need to investigate this further. :)

**Comment by dennisdoomen** [comment]
Good thinking Jakub. It seems I've been a bit on the naive side with this feature request. 

With respect to the `WithSubjectTypeFrom`, would it be possible to use the `nameof` operator instead? I'm not sure it works on anonymous types. 

Do you think we can have two ways of specifying the subject's type? Maybe we can have a `Map<TSubject>` _and_ a `WithSubjectType<TSubject>().Map()`

**Comment by nohwnd** [comment]
> Good thinking Jakub. It seems I've been a bit on the naive side with this feature request.

Don' worry, I think we both were :) 

> With respect to the WithSubjectTypeFrom, would it be possible to use the nameof operator instead? I'm not sure it works on anonymous types.

I don't think `nameof` will help us here, if you did `nameof(subject)` you would get "subject" string, and even if it returned the type (as in subject.GetType()) it would not help us. The purpose of the `WithSubjectType*` method is to fill in the `TSubject` type parameter with a concrete type so we can later define the expressions in `Map(Expression<Func<TSubject, object>> subjectExpression, Expression<Func<TExpectation, object>> expectationExpression)`. This all has to happen on type-check time, so all we can do is either specify the types as type parameters, or use strings to specify the mapping paths.

> Do you think we can have two ways of specifying the subject's type? Maybe we can have a Map<TSubject> and a WithSubjectType<TSubject>().Map()

That totally should be possible, we already know the type of `TExpectation` from the `EquivalenceAssertionOptions<TExpectation>` so we just need to specify one more type parameter for `TSubject`, but the same approach with providing the instance of subject to infer it's anonymous type would have to apply. 

I would wait with shaping the fluent api till the end, or maybe till next minor release, so people have chance to find issues with a single approach, even though it means a slightly more typing, and you don't have to intorduce confusing api that will later need to be redefined with breaking changes.

**Comment by dennisdoomen** [comment]
> I don't think nameof will help us here, if you did nameof(subject)

You're misunderstanding me. I'm suggesting an extra overload of `Map` that takes the name of the subject's property and an expression to select the expectation's property to map. I've just checked, and you can use the `nameof` operator on a property of an anonymous type. It would allow you to do this:

```csharp
subject.Should().
    BeEquivalentTo(other, options => options
       .Including(p=>p.FirstName));
       .Map(nameof(subject.Name), o => o.FirstName)
```

**Comment by nohwnd** [comment]
Gotcha, but `nameof` would only work for top-level members, for child members you have to fallback to strings. The `Expression<Func<T, object>>` is in the end also just a type safe way of providing a string, so we should be able to expose the mapping as expressions, and as strings. I don't want to promise it just yet, I need to see first if I need to know the type of the `TSubject` somewhere to avoid ambiguous mappings. But on a first look that should be totally possible.

**Comment by MateuszStepniak** [comment]
This is such a common task while writing miscroservices that it definitely would be awesome to have this feature.

**Comment by tparikka** [comment]
I'm currently trying to run assertions against the response of a .NET Core API, where the DTO has a property of `Description` (that is being projected to via AutoMapper at an assembly level with reflection) but the property is decorated with a `JsonPropertyName` attribute from `System.Text.Json` that causes it to return as `Value` in the API response. Being able to map properties would be extremely valuable to me. For now I'll have to work around it by doing my own projection in my test assembly.

**Comment by SubjectiveReality** [comment]
👍 this would be a massive timesaver -- let us know if we can contribute...

**Comment by dennisdoomen** [comment]
I have a couple of options that I'm considering, each supporting using both expressions as well as a textual path.

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingSubjectMapping<ClassWithSomeFieldsAndProperties>(map => map
        .Add(e => e.Property2, s => s.Property1)
        .Add("Propery2", "Property1")
    ));

subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingMappingTo<ClassWithSomeFieldsAndProperties>(e => e.Property2, s => s.Property1)
    .UsingMapping("Propery2", "Property1"));

subject.Should().BeEquivalentTo(expectation, opt => opt
    .MappingFrom(e => e.Property2).To<ClassWithSomeFieldsAndProperties>(s => s.Property1)
    .MappingFrom("Propery2").To("Property1"));
```

Any opinions or suggestions?

**Comment by MateuszStepniak** [comment]
1. There is an overhead of additional Add operation when there is only one property to map. Still, you only need to invoke concise Add with each next property. That's definitely an advantage. The more properties to map, the best return on investment we have here. The question is which scenario is more common in the real world.
2. Just one call to do the single property mapping. Sweet in this scenario. Worse when having more properties.
3. The closest to natural language, which is a plus. Minus: requires two method calls. 

Further thoughts:

1
This is like a builder, that makes complex scenarios easier, but adds overhead in simpler ones. From my experience, those scenarios that require more properties to map are not that uncommon. And this is so tiring to type the same stuff over and over again then. This version would protect us from such things happening and all the unnecessary boilerplate.

2, 3
Those shine in simpler mappings. 

2 vs 3
I always wonder, do I prefer names as close to natural language as possible since its processing is so much easier for our minds and the boost from it is huge. Then I would pick number 3.
And then I think that after time you might get used to more symbolic expressions and naming conventions. Similarly that you can get used to and learn reading mathematical expressions and describe what you mean in a faster more concise way thanks to them. In this case, I would pick number 2.  But still, in both numbers 1 and 2, the name of the (first) method is pretty long.

**Comment by lg2de** [comment]
The more fluent the better...
I prefer a format close to natural language. For me version 3 is most close.
BUT, what happens if the call to `To` is omit? What is the behavior? Can the compiler detects the missing call?
If so V3 looks like my favorite.

If not I prefer V2 because I cannot identify the advantages of V1 over V2.
With both it should be possible to specify one or more mappings.
Then I prefer the less complex version.

**Comment by abatishchev** [comment]
I'd personally prefer options 1 or 2.

**Comment by jnyrup** [comment]
I like option 2 the most. 
Starring at at the proposals for two minutes, it looks like option 2 will also have the simplest implementation. 

I recall that the Using+When combo have some difficulties.
E.g. remembering to call When. 
That's my thing against options 3 even though it might read closer to natural language.

**Comment by michal-ciechan** [comment]
3 gets my vote as it reads best

But if 2 is much simpler to implement/maintain I would go with that. 

Some suggestions to 2 would be:

1. Use `Mapping` instead, seems short intuitive enough

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .Mapping<ClassWithSomeFieldsAndProperties>(e => e.Property2, s => s.Property1)
    .Mapping("Propery2", "Property1")
);
```

Failing that, maybe use `WithMapping(...)` as we already have about 10 `WithXxxx` methods?

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .WithMapping<ClassWithSomeFieldsAndProperties>(e => e.Property2, s => s.Property1)
    .WithMapping("Propery2", "Property1")
);
```

2. Order being `Source` -> `Expectation` (from your snippet, based on anonymous parameter names)

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingMappingTo<SomeInnerSubjectClass, Some>(s => s.Property1, e => e.Property2)
    .UsingMappingTo("Propery1", "Property2")
);
```

### Harder / Out of scope features that could come after

3. Not needing to specify the type, having `Should().BeEquivalentTo(expected)` return a custom generic type which has knowledge of both Source + Expected type:

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingMappingTo(e => e.Property2, s => s.Property1)
);
```

4. Allow specifying any 2 types to use (which could exist at any level):

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingMappingTo("Inner.Propery1", "Inner.Property2")
);
```

5. Allow specifying any 2 types to use (which could exist at any level):

```csharp
subject.Should().BeEquivalentTo(expectation, opt => opt
    .UsingMappingTo<SomeInnerSubjectClass, SomeInnerExpectationClass>(e => e.Property2, s => s.Property1)
);
```

**Comment by dennisdoomen** [comment]
```
subject.Should().BeEquivalentTo(expectation, opt => opt
    .WithMapping<ClassWithSomeFieldsAndProperties>(e => e.Property2, s => s.Property1)
    .WithMapping("Propery2", "Property1")
);
```

I like this one. It's the one I'll try to implement.

**Comment by FBryant87** [comment]
Exciting, we were trying to find if this was possible today!

**Comment by dennisdoomen** [comment]
While trying to implement this, I realized we also want to be able to map properties of nested objects, which type can be different. 

For mapping properties on the root object, we only need to specify the type of the subject if you want to use expressions.
```
.WithMapping<Subject>(e => e.Property2, s => s.Property1);  
.WithMapping("Propery2", "Property1");
```

For mapping properties on the root object, unless you specify a nested path, you always need to specify the type of the subject and the expectation.
```
.WithMapping<Expectation, Subject>(e => e.Property2, s => s.Property1);  
.WithMapping<Expectation, Subject>("Propery2", "Property1");
.WithMapping("PropertyToGetToExpecation.Propery2", "PropertyToGetToSubject.Property1");
```

**Comment by dennisdoomen** [comment]
Here are some of the tests I'm using to drive the implementation:

https://github.com/fluentassertions/fluentassertions/pull/1742/files#diff-53e5ae72194305b58e8010fc0c53e6bd5bd38e636787ab512c659f3db41fb3f4R3434-R3589

**Comment by dennisdoomen** [comment]
Will be part of 6.5.0

**Comment by MateuszStepniak** [comment]
That's splendid news, thank you for your work!

**Comment by lorisp1** [comment]
Hello, I am facing an issue similar to the `x => x.State.Code, x => x.State` one from @fredgate and I found this issue. It is not clear to me whether #1742 solves that case (I'm using FA 6.12.2). That is, I used `WithMapping` to map properties with same nesting and types but different names a lot of times, but now I need to check the equality between two properties in the same way fredgate proposed and I am not being able to obtain the desired behaviour.
Am I missing something?
Thanks
